src/transaction/validation.rs

```diff
@@ -86,49 +171,64 @@ impl Transaction {
             Err(ValidationError::TransactionWitnessesMax)?
         }
 
-        let input_colors: Vec<&Color> = self.input_colors().collect();
-        for input_color in input_colors.as_slice() {
+        self.input_colors_unique().try_for_each(|input_color| {
             // check for duplicate change outputs
             if self
                 .outputs()
                 .iter()
                 .filter_map(|output| match output {
-                    Output::Change { color, .. } if input_color == &color => Some(()),
+                    Output::Change { color, .. } if input_color == color => Some(()),
+                    Output::Change { color, .. }
+                        if color != &Color::default() && input_color == color =>
+                    {
+                        Some(())
+                    }
                     _ => None,
                 })
                 .count()
                 > 1
             {
-                Err(ValidationError::TransactionOutputChangeColorDuplicated)?
+                return Err(ValidationError::TransactionOutputChangeColorDuplicated);
             }
 
             // check for duplicate variable outputs
             if self
                 .outputs()
                 .iter()
                 .filter_map(|output| match output {
-                    Output::Variable { color, .. } if input_color == &color => Some(()),
+                    Output::Variable { color, .. } if input_color == color => Some(()),
                     _ => None,
                 })
                 .count()
                 > 1
             {
-                Err(ValidationError::TransactionOutputVariableColorDuplicated)?
+                return Err(ValidationError::TransactionOutputVariableColorDuplicated);
```

Comment:
Unrelated to this PR, but I think this validation rule may be incorrect. 

If there are multiple TRO calls of the same asset id, you'd need to have duplicate colors since each TRO can only set a variable output once.  Also, users with variable outputs shouldn't need to set to color in advance anyways since it will be zeroed out prior to execution.

cc: @adlerjohn

Reply:
https://github.com/FuelLabs/fuel-tx/issues/82